pkg/reconciler/route/reconciler.go

```diff
@@ -0,0 +1,177 @@
+	*reconciler.Base
+
+	// listers index properties about resources
+	routeLister          kflisters.RouteLister
+	virtualServiceLister istiolisters.VirtualServiceLister
+}
+
+// Check that our Reconciler implements controller.Reconciler
+var _ controller.Reconciler = (*Reconciler)(nil)
+
+// Reconcile is called by Kubernetes.
+func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
+	logger := logging.FromContext(ctx)
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	original, err := r.routeLister.Routes(namespace).Get(name)
+	switch {
+	case apierrs.IsNotFound(err):
+		logger.Errorf("Route %q no longer exists\n", name)
+		return nil
+
+	case err != nil:
+		return err
+
+	case original.GetDeletionTimestamp() != nil:
+		return nil
+	}
+
+	// Don't modify the informers copy
+	toReconcile := original.DeepCopy()
+
+	// Reconcile this copy of the service and then write back any status
+	// updates regardless of whether the reconciliation errored out.
+	reconcileErr := r.ApplyChanges(ctx, toReconcile)
+	if equality.Semantic.DeepEqual(original.Status, toReconcile.Status) {
+		// If we didn't change anything then don't call updateStatus.
+		// This is important because the copy we loaded from the informer's
+		// cache may be stale and we don't want to overwrite a prior update
+		// to status with this stale state.
+
+	} else if _, uErr := r.updateStatus(toReconcile); uErr != nil {
+		logger.Warnw("Failed to update Route status", zap.Error(uErr))
+		return uErr
+	}
+
+	return reconcileErr
+}
+
+// ApplyChanges updates the linked resources in the cluster with the current
+// status of the Route .
+func (r *Reconciler) ApplyChanges(ctx context.Context, route *v1alpha1.Route) error {
+	logger := logging.FromContext(ctx)
+
+	route.Status.InitializeConditions()
+	virtualServiceName := resources.VirtualServiceName(route.Spec.Hostname, route.Spec.Domain, route.Spec.Path)
+
+	// Sync VirtualService
+	{
+		desired, err := resources.MakeVirtualService(route)
+		if err != nil {
+			return err
+		}
+
+		actual, err := r.virtualServiceLister.VirtualServices(desired.GetNamespace()).Get(desired.Name)
+		if errors.IsNotFound(err) {
+			// VirtualService doesn't exist, make one.
+			actual, err = r.SharedClientSet.Networking().VirtualServices(desired.GetNamespace()).Create(desired)
+			if err != nil {
+				return err
+			}
+		} else if err != nil {
+			return err
+		} else if !metav1.IsControlledBy(actual, route) {
+			route.Status.MarkVirtualServiceNotOwned(virtualServiceName)
+			return fmt.Errorf("route: %q does not own VirtualService: %q", route.Name, virtualServiceName)
+		} else if actual, err = r.reconcile(desired, actual); err != nil {
+			return err
+		}
+
+		route.Status.PropagateVirtualServiceStatus(actual)
+	}
+
+	// If the VirtualService isn't ready (either it's coming up or being
+	// deleted) then this reconciliation process can't continue until we get
+	// notified it is.
```

Comment:
Doesn't seem like this is necessary now. Will it become necessary in later PRs?

Reply:
I mostly have it there for parity with the space controller. I can remove it, but I don't think its hurting anything. WDYT?

Reply:
Let's remove it now, and readd it later if the route CRD really needs this hard check. Readding it should be easy, since we can copy the space reconcilier